### PR TITLE
Meta object

### PR DESCRIPTION
**Pull request by delvedor**
First part of the work started from the discussion in https://github.com/fastify/fastify/issues/456.
Once the core is ready we will merge and release this as well.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,6 +36,24 @@ module.exports = fp(function (fastify, opts, next) {
...
+  // your plugin code
+  next()
+}
+
+module.exports = fp(plugin, {
+  version: '0.x',
+  dependencies: {
+    fastify: ['plugin1', 'plugin2'],
+    reply: ['compress']
+  }
```
Comment (mcollina) [inline README.md]: I'm a bit lost on how this dependencies block would be used.

File: index.js [hunk 2]
```diff
@@ -3,30 +3,43 @@
 const semver = require('semver')
 const console = require('console')
 
-function plugin (fn, version) {
+function plugin (fn, options) {
```
Comment (mcollina) [inline index.js]: I would keep it backward-compatible.

Comment (delvedor) [reply index.js]: It is https://github.com/fastify/fastify-plugin/blob/88997b05faf79c3a4d9b777d4d474b0f3eeb83a1/index.js#L15

File: README.md (hunk 1 above)
Comment (delvedor) [reply README.md]: Is still a WIP, checkout https://github.com/fastify/fastify/pull/466.

File: README.md [hunk 3]
```diff
@@ -9,9 +9,10 @@ When you build plugins for Fastify and you want that them to be accessible in th
 2. Use this module
 
 #### Usage
-`fastify-plugin` can do two things for you:
+`fastify-plugin` can do three things for you:
 - Add the `skip-override` hidden property
 - Check the bare-minimum version of Fastify
+- Ass some custom meta to the plugin
```
Comment (jsumners) [inline README.md]: How about we lose the "Ass" and go with this instead:

> Asses plugin metadata

File: README.md [hunk 4]
```diff
@@ -35,6 +36,24 @@ module.exports = fp(function (fastify, opts, next) {
 
 You can check [here](https://github.com/npm/node-semver#ranges) how to define a `semver` range.
 
+You can also pass some metadata thatr will be handled by Fastify, such as the dependencies of your plugin.
```
Comment (jsumners) [inline README.md]: Strike "thatr" and replace with "that".

File: README.md (hunk 3 above)
Comment (delvedor) [reply README.md]: Oooops :D

File: test.js [hunk 5]
```diff
@@ -67,10 +67,10 @@ test('should throw if the version number is not a string', t => {
   t.plan(1)
 
   try {
-    fp(() => {}, 12)
+    fp(() => {}, { version: 12 })
```
Comment (allevo) [inline test.js]: Can we keep this test? Or this PR is a breaking change?

Comment (delvedor) [reply test.js]: At the moment the second parameter should be a string or object.

File: index.js (hunk 2 above)
Comment (mcollina) [reply index.js]: this is not true anymore, correct?

Comment (delvedor) [reply index.js]: Correct.

File: README.md [hunk 6]
```diff
@@ -8,11 +8,13 @@ When you build plugins for Fastify and you want that them to be accessible in th
 1. Use the `skip-override` hidden property
 2. Use this module
 
+In addition if you use this module when creating new plugins, you can declare the dependencies, the name and the expected Fastify version that your plugins needs.
```
Comment (jsumners) [inline README.md]: `plugins needs` => `plugin needs`

File: README.md [hunk 7]
```diff
@@ -8,11 +8,13 @@ When you build plugins for Fastify and you want that them to be accessible in th
...
+
 #### Usage
 `fastify-plugin` can do three things for you:
 - Add the `skip-override` hidden property
 - Check the bare-minimum version of Fastify
-- Pass some custom meta to the plugin
+- Pass some custom meta of the plugin to Fastify
```
Comment (jsumners) [inline README.md]: `meta` => `metadata`

**Pull request merged**